src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,265 @@
+#endif
+
+static void NetSecurityNative_HandleError(uint32_t majorStatus, GssBuffer** outBufferHandle, uint32_t* outBufferLength)
+{
+    assert(outBufferHandle != nullptr);
+    assert(outBufferLength != nullptr);
+    if (GSS_ERROR(majorStatus) || (*outBufferHandle == nullptr) ||
+        (*outBufferHandle)->length > UnsignedCast(std::numeric_limits<uint32_t>::max()))
+    {
+        uint32_t relBufferStatus;
+        gss_release_buffer(&relBufferStatus, *outBufferHandle);
+        *outBufferHandle = nullptr;
+        *outBufferLength = 0;
+    }
+    else
+    {
+        *outBufferLength = static_cast<uint32_t>((*outBufferHandle)->length);
+    }
+}
+
+extern "C" uint32_t NetSecurityNative_AcquireCredSpNego(uint32_t* minorStatus,
+                                                        GssName* desiredName,
+                                                        int32_t isInitiate,
+                                                        GssCredId** outputCredHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(outputCredHandle != nullptr);
+    assert(isInitiate == 0 || isInitiate == 1);
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID_set_desc gss_mech_spnego_OID_set_desc = {.count = 1, .elements = GSS_SPNEGO_MECHANISM};
+#else
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .elements = static_cast<void*>(gss_mech_value)};
+    gss_OID_set_desc gss_mech_spnego_OID_set_desc = {.count = 1, .elements = &gss_mech_spnego_OID_desc};
+#endif
+    gss_cred_usage_t credUsage = isInitiate ? GSS_C_INITIATE : GSS_C_ACCEPT;
+    return gss_acquire_cred(
+        minorStatus, desiredName, 0, &gss_mech_spnego_OID_set_desc, credUsage, outputCredHandle, nullptr, nullptr);
+}
+
+extern "C" uint32_t NetSecurityNative_DeleteSecContext(uint32_t* minorStatus, GssCtxId** contextHandle)
+{
+    assert(contextHandle != nullptr);
+    assert(minorStatus != nullptr);
+    return gss_delete_sec_context(minorStatus, contextHandle, GSS_C_NO_BUFFER);
+}
+
+extern "C" uint32_t NetSecurityNative_DisplayStatus(uint32_t* minorStatus,
+                                                    uint32_t statusValue,
+                                                    int32_t isGssMechCode,
+                                                    GssBuffer** outBufferHandle,
+                                                    uint32_t* statusLength)
+{
+    assert(minorStatus != nullptr);
+    assert(outBufferHandle != nullptr);
+    assert(statusLength != nullptr);
+    assert(isGssMechCode == 0 || isGssMechCode == 1);
+    int statusType = isGssMechCode ? GSS_C_MECH_CODE : GSS_C_GSS_CODE;
+    *outBufferHandle = new GssBuffer();
+    uint32_t majorStatus =
+        gss_display_status(minorStatus, statusValue, statusType, GSS_C_NO_OID, nullptr, *outBufferHandle);
+    NetSecurityNative_HandleError(majorStatus, outBufferHandle, statusLength);
+    return majorStatus;
```

Comment:
What happens if majorStatus is success but the checks in HandleError see, for example, that the length is too long, and end up releasing and nulling out the buffer?  We then still return the success status?